mlir/lib/Dialect/EmitC/Transforms/AddReflectionMap.cpp

```diff
@@ -0,0 +1,180 @@
+    rewriter.setInsertionPointToStart(funcBody);
+
+    // Collect all field names
+    std::vector<std::pair<std::string, std::string>> fieldNames;
+    classOp.walk([&](mlir::emitc::FieldOp fieldOp) {
+      if (mlir::Attribute attrsAttr =
+              fieldOp->getAttrDictionary().get("attrs")) {
+        if (DictionaryAttr innerDictAttr =
+                dyn_cast<mlir::DictionaryAttr>(attrsAttr)) {
+          ArrayAttr arrayAttr = dyn_cast<mlir::ArrayAttr>(
+              innerDictAttr.getNamed(attributeName)->getValue());
+          if (!arrayAttr.empty()) {
+            StringAttr stringAttr = dyn_cast<mlir::StringAttr>(arrayAttr[0]);
+            std::string indexPath = stringAttr.getValue().str();
+            fieldNames.emplace_back(indexPath, fieldOp.getName().str());
+          }
+          if (arrayAttr.size() > 1) {
+            fieldOp.emitError() << attributeName
+                                << " attribute must "
+                                   "contain at most one value, but found "
+                                << arrayAttr.size() << " values.";
+            return;
+          }
+        }
+      }
+    });
+
+    std::string mapInitializer = "{ ";
+    for (size_t i = 0; i < fieldNames.size(); ++i) {
+      mapInitializer += " { \"" + fieldNames[i].first + "\", " +
+                        "reinterpret_cast<char*>(&" + fieldNames[i].second +
+                        ")",
+          mapInitializer += " }";
+      if (i < fieldNames.size() - 1)
+        mapInitializer += ", ";
+    }
+    mapInitializer += " }";
+
+    emitc::OpaqueType iteratorType = mlir::emitc::OpaqueType::get(
+        context, "std::map<std::string, char*>::const_iterator");
+
+    emitc::ConstantOp bufferMap = rewriter.create<emitc::ConstantOp>(
+        classOp.getLoc(), mapType,
+        emitc::OpaqueAttr::get(context, mapInitializer));
+
+    mlir::Value nameArg = getBufferFunc.getArgument(0);
+    emitc::CallOpaqueOp it = rewriter.create<emitc::CallOpaqueOp>(
+        classOp.getLoc(), iteratorType, rewriter.getStringAttr("find"),
+        mlir::ValueRange{bufferMap.getResult(), nameArg});
+    emitc::CallOpaqueOp endIt = rewriter.create<emitc::CallOpaqueOp>(
+        classOp.getLoc(), iteratorType, rewriter.getStringAttr("end"),
+        bufferMap.getResult());
+    emitc::CallOpaqueOp isEnd = rewriter.create<emitc::CallOpaqueOp>(
+        classOp.getLoc(), rewriter.getI1Type(),
+        "operator==", mlir::ValueRange{it.getResult(0), endIt.getResult(0)});
+    emitc::ConstantOp nullPtr = rewriter.create<emitc::ConstantOp>(
+        classOp.getLoc(), charType, emitc::OpaqueAttr::get(context, "nullptr"));
+    emitc::CallOpaqueOp second = rewriter.create<emitc::CallOpaqueOp>(
+        classOp.getLoc(), charType, "second", it.getResult(0));
+
+    emitc::ConditionalOp result = rewriter.create<emitc::ConditionalOp>(
+        classOp.getLoc(), charType, isEnd.getResult(0), nullPtr.getResult(),
+        second.getResult(0));
```

Comment:
For blocks like this, where you're basically spelling out the C++ code, you may want to write that down in a comment, so its easy to see what operations you're doing.